File: Core/src/main/java/dev/lynith/core/ui/screen/MainMenu.java
package dev.lynith.core.ui.screen;

import dev.lynith.core.ui.Screen;
import dev.lynith.core.ui.callbacks.impl.MouseClick;
import dev.lynith.core.ui.components.Button;
import dev.lynith.core.ui.styles.ComponentStyles;
import dev.lynith.core.ui.styles.Spacing;
import dev.lynith.core.utils.GuiScreens;

public class MainMenu extends Screen<MainMenu, ComponentStyles<MainMenu>> {

    public MainMenu() {
        super(GuiScreens.MAIN_MENU);
    }

    @Override
    public void init() {
        style(styles -> {
            styles.setPadding(Spacing.from(10));
        });

        add(
            new Button("Singleplayer")
                .position(0, 0)
                .size(100, 20)
                .listener(MouseClick.class, (mouseX, mouseY) ->
                    bridge().getRenderer().setCurrentScreen(GuiScreens.SINGLEPLAYER_SELECT, GuiScreens.MAIN_MENU)
                ),
            new Button("Multiplayer")
                .position(0, 30)
                .size(100, 20)
                .listener(MouseClick.class, (mouseX, mouseY) ->
                    bridge().getRenderer().setCurrentScreen(GuiScreens.MULTIPLAYER_SELECT, GuiScreens.MAIN_MENU)
                ),
            new Button("Options")
                .position(0, 60)
                .size(100, 20)
                .listener(MouseClick.class, (mouseX, mouseY) ->
                    bridge().getRenderer().setCurrentScreen(GuiScreens.OPTIONS, GuiScreens.MAIN_MENU)
                ),
            new Button("Quit")
                .position(0, 90)
                .size(100, 20)
                .listener(MouseClick.class, (mouseX, mouseY) ->
                    bridge().getGame().shutdown()
                )
        );

        super.init();
    }
}

File: Core/src/main/kotlin/dev/lynith/core/ui/screens/MainMenu.kt
package dev.lynith.core.ui.screens

import dev.lynith.core.Platform
import dev.lynith.core.hud.HudConfigScreen
import dev.lynith.core.ui.BoundingBox
import dev.lynith.core.ui.components.Screen
import dev.lynith.core.ui.callbacks.impl.Clicked
import dev.lynith.core.ui.components.impl.Block
import dev.lynith.core.ui.components.impl.Button
import dev.lynith.core.ui.components.impl.Label
import dev.lynith.core.ui.layouts.LayoutProperties
import dev.lynith.core.ui.nvg.Font
import dev.lynith.core.ui.styles.impl.Color
import dev.lynith.core.ui.styles.impl.Position
import dev.lynith.core.ui.units.px

class MainMenu : Screen() {

    override fun render(mouseX: Int, mouseY: Int, delta: Float) {
        background()
    }

    override fun init() {
        layout {
            direction = LayoutProperties.Direction.Vertical
            align = LayoutProperties.Align.Center
            justify = LayoutProperties.Justify.Center
        }

        children (
            Button().configure {
                style {
                    position = Position(Position.PositionType.ABSOLUTE)
                    backgroundColor = Color(0, 255, 0)
                }

                hoverStyles {
                    backgroundColor = Color(0, 0, 255)
                }

                bounds = BoundingBox(
                    x = 10.px,
                    y = 10.px,
                )

                text = "Test"

                on<Clicked> {
                    Platform.renderer.setScreen(HudConfigScreen())
                }
            },

            Block().configure {
                layout {
                    direction = LayoutProperties.Direction.Vertical
                    align = LayoutProperties.Align.Center
                    justify = LayoutProperties.Justify.Center
                    childWidth = LayoutProperties.ChildSize.Fill
                    gap = LayoutProperties.Gap(
                        y = 10.px
                    )
                }

                bounds = BoundingBox(
                    width = 300.px
                )

                children(
                    Label().configure {
                        text = "ZABU"

                        style {
                            font.change {
                                size = 56.px
                                weight = Font.FontWeight.BOLD
                                horizontalAlign = Font.HorizontalAlign.CENTER
                                letterSpacing = 5.px
                            }
                        }
                    },

                    Button().configure {
                        text = "Singleplayer"

                        on<Clicked> {
                            Platform.renderer.displaySingleplayerSelectorScreen()
                        }
                    },

                    Button().configure {
                        text = "Multiplayer"

                        on<Clicked> {
                            Platform.renderer.displayMultiplayerSelectorScreen()
                        }
                    },

                    Block().configure {
                        layout {
                            direction = LayoutProperties.Direction.Horizontal
                            align = LayoutProperties.Align.Center
                            justify = LayoutProperties.Justify.Center
                            childWidth = LayoutProperties.ChildSize.Fill
                            gap = LayoutProperties.Gap(
                                x = 10.px
                            )
                        }

                        children(
                            Button().configure {
                                text = "Options"

                                on<Clicked> {
                                    Platform.renderer.displayOptionsScreen()
                                }
                            },

                            Button().configure {
                                text = "Quit"

                                on<Clicked> {
                                    Platform.minecraft.scheduleStop()
                                }
                            }
                        )
                    }
                )
            }
        )
    }
}
